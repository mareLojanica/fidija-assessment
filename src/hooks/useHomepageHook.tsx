import { useContext } from "react";
import { HomepageContext } from "../context/HomepageContext";
import { HomepageState } from "../types/poviders.types";

const useHomepage = (): HomepageState => {
  const context = useContext(HomepageContext);
  if (!context) {
    throw new Error("useHomepage must be used within a HomepageProvider");
  }
  return context;
};

export default useHomepage;
