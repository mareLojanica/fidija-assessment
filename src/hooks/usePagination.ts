import { useState, useMemo } from "react"

interface UsePaginationProps<T> {
	data: T[]
	itemsPerPage: number
}

export const usePagination = <T>({
	data,
	itemsPerPage,
}: UsePaginationProps<T>) => {
	const [currentPage, setCurrentPage] = useState(0)

	const pageCount = useMemo(
		() => Math.ceil(data.length / itemsPerPage),
		[data, itemsPerPage]
	)

	const paginatedData = useMemo(
		() =>
			data.slice(
				currentPage * itemsPerPage,
				(currentPage + 1) * itemsPerPage
			),
		[data, currentPage, itemsPerPage]
	)

	const goToNextPage = () =>
		setCurrentPage((prev) => Math.min(prev + 1, pageCount - 1))
	const goToPrevPage = () => setCurrentPage((prev) => Math.max(prev - 1, 0))
	const goToPage = (page: number) => setCurrentPage(page)

	return {
		paginatedData,
		currentPage,
		pageCount,
		goToNextPage,
		goToPrevPage,
		goToPage,
		setCurrentPage,
	}
}
