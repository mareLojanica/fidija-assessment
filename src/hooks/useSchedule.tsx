import { useQuery } from "@tanstack/react-query"
import { fetchSchedule } from "../api/api"
import { SCHEDULE_QUERY_KEY } from "../constants"

export const useSchedule = (selectedDate: string) => {
	return useQuery({
		queryKey: [SCHEDULE_QUERY_KEY, selectedDate],
		queryFn: () => fetchSchedule(selectedDate),
	})
}
