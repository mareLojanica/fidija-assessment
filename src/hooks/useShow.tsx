import { useQuery } from "@tanstack/react-query"
import { fetchShow } from "../api/api"
import { SHOW_QUERY_KEY } from "../constants"

export const useShow = (showId: string) => {
	return useQuery({
		queryKey: [SHOW_QUERY_KEY, showId],
		queryFn: () => fetchShow(showId),
	})
}
