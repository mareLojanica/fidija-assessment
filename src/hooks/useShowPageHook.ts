import { useContext } from "react";
import { ShowPageContext } from "../context/ShowPageContext";

export const useShowPage = () => {
  const context = useContext(ShowPageContext);
  if (!context) {
    throw new Error("useShowPage must be used within a ShowPageProvider");
  }
  return context;
};

export default useShowPage;
