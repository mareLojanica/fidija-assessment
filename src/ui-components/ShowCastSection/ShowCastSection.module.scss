.cast-list {
  margin-top: 2rem;

  &__title {
    font-size: 20px;
    font-weight: bold;
  }

  &__content {
    display: flex;
    flex-direction: column;
    gap: 10px;
  }

  &__item {
    display: flex;
    align-items: center;
    gap: 15px;
    padding: 15px 0;
    border-bottom: 1px solid #bfbfbf;
    height: 80px;

    @media (max-width: 768px) {
      border-bottom: none;
    }
  }

  &__img-wrapper {
    width: 50px;
    height: 50px;
    border-radius: 50%;
    background-color: #bfbfbf;
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &__img {
    height: 100%;
    width: 100%;
    border-radius: 50%;
    object-fit: contain;
  }

  &__details {
    display: flex;
    flex-direction: column;
  }

  &__name {
    font-weight: bold;
  }

  &__character {
    font-size: 12px;
    color: gray;
  }

  &__empty {
    font-size: 14px;
    color: #888;
  }
}
