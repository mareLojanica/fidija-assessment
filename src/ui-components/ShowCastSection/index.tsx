import React from "react";
import styles from "./ShowCastSection.module.scss";
import Pagination from "../../ui-components/Pagination";
import { ShowCastSectionProps } from "../../types/ui-component.types";
import ShowSection from "../ShowSection";
import { usePagination } from "../../hooks/usePagination";
import CastListItem from "../CastListItem";
import { ACTORS_PER_PAGE } from "../../constants";

const ShowCastSection: React.FC<ShowCastSectionProps> = ({ cast }) => {
  const { paginatedData, currentPage, pageCount, goToPage } = usePagination({
    data: cast,
    itemsPerPage: ACTORS_PER_PAGE,
  });

  return (
    <ShowSection title="Starring">
      <article data-testid="show-cast-section">
        {paginatedData.length > 0 ? (
          paginatedData.map((actor) => (
            <React.Fragment key={actor.person.id}>
              <CastListItem
                actor={{
                  person: {
                    id: actor.person.id,
                    name: actor.person.name,
                    image: actor.person.image,
                  },
                  character: {
                    name: actor.character.name,
                  },
                }}
              />
            </React.Fragment>
          ))
        ) : (
          <p className={styles["cast-list__empty"]}>
            No cast information available.
          </p>
        )}
        {pageCount > 1 && (
          <Pagination
            pageCount={pageCount}
            currentPage={currentPage}
            onPageChange={goToPage}
          />
        )}
      </article>
    </ShowSection>
  );
};

export default ShowCastSection;
