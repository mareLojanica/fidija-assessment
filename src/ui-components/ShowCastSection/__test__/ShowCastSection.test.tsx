import React from "react";
import { render, screen } from "@testing-library/react";
import ShowCastSection from "../";
import "@testing-library/jest-dom";
import { usePagination } from "../../../hooks/usePagination";
jest.mock("../../../hooks/usePagination", () => ({
  usePagination: jest.fn(),
}));

jest.mock("../../CastListItem", () =>
  jest.fn(({ actor }) => (
    <div data-testid="cast-list-item">
      <p>{actor.person.name}</p>
      <p>{actor.character.name}</p>
    </div>
  ))
);

jest.mock("../../Pagination", () =>
  jest.fn(({ currentPage, pageCount, onPageChange }) => (
    <div data-testid="pagination">
      <button
        disabled={currentPage === 1}
        onClick={() => onPageChange(currentPage - 1)}
      >
        Prev
      </button>
      <button
        disabled={currentPage === pageCount}
        onClick={() => onPageChange(currentPage + 1)}
      >
        Next
      </button>
    </div>
  ))
);

describe("ShowCastSection Component", () => {
  const mockCast = [
    {
      person: {
        id: 1,
        name: "Bryan Cranston",
        url: "https://example.com/bryan",
        country: { name: "USA", code: "US", timezone: "America/New_York" },
        birthday: "1956-03-07",
        deathday: null,
        gender: "Male",
        image: {
          medium: "https://example.com/bryan-medium.jpg",
          original: "https://example.com/bryan.jpg",
        },
        updated: 1678900000,
        _links: { self: { href: "https://example.com/bryan" } },
      },
      character: {
        id: 101,
        url: "https://example.com/walter-white",
        name: "Walter White",
        image: {
          medium: "https://example.com/walter-medium.jpg",
          original: "https://example.com/walter.jpg",
        },
        _links: { self: { href: "https://example.com/walter-white" } },
      },
      self: false,
      voice: false,
    },
    {
      person: {
        id: 2,
        name: "Aaron Paul",
        url: "https://example.com/aaron",
        country: { name: "USA", code: "US", timezone: "America/Los_Angeles" },
        birthday: "1979-08-27",
        deathday: null,
        gender: "Male",
        image: {
          medium: "https://example.com/aaron-medium.jpg",
          original: "https://example.com/aaron.jpg",
        },
        updated: 1678900001,
        _links: { self: { href: "https://example.com/aaron" } },
      },
      character: {
        id: 102,
        url: "https://example.com/jesse-pinkman",
        name: "Jesse Pinkman",
        image: {
          medium: "https://example.com/jesse-medium.jpg",
          original: "https://example.com/jesse.jpg",
        },
        _links: { self: { href: "https://example.com/jesse-pinkman" } },
      },
      self: false,
      voice: false,
    },
  ];

  beforeEach(() => {
    (usePagination as jest.Mock).mockReturnValue({
      paginatedData: mockCast,
      currentPage: 1,
      pageCount: 2,
      goToNextPage: jest.fn(),
      goToPrevPage: jest.fn(),
      goToPage: jest.fn(),
      setCurrentPage: jest.fn(),
    });
  });

  test("renders cast section title", () => {
    render(<ShowCastSection cast={mockCast} />);

    expect(screen.getAllByText("Starring")[0]).toBeInTheDocument();
  });

  test("renders cast list when data is available", () => {
    render(<ShowCastSection cast={mockCast} />);

    expect(screen.getAllByTestId("cast-list-item")).toHaveLength(2);
    expect(screen.getByText("Bryan Cranston")).toBeInTheDocument();
    expect(screen.getByText("Walter White")).toBeInTheDocument();
    expect(screen.getByText("Aaron Paul")).toBeInTheDocument();
    expect(screen.getByText("Jesse Pinkman")).toBeInTheDocument();
  });

  test("renders 'No cast information available' when cast list is empty", () => {
    (usePagination as jest.Mock).mockReturnValue({
      paginatedData: [],
      currentPage: 1,
      pageCount: 1,
      goToNextPage: jest.fn(),
      goToPrevPage: jest.fn(),
      goToPage: jest.fn(),
      setCurrentPage: jest.fn(),
    });

    render(<ShowCastSection cast={[]} />);

    expect(
      screen.getByText("No cast information available.")
    ).toBeInTheDocument();
  });

  test("renders pagination when more than one page exists", () => {
    render(<ShowCastSection cast={mockCast} />);

    expect(screen.getByTestId("pagination")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<ShowCastSection cast={mockCast} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
