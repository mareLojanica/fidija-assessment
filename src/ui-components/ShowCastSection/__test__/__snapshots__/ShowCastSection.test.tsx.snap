// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowCastSection Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="show-info"
  >
    <h2
      class="show-info__title"
    >
      Starring
    </h2>
    <div
      class="show-info__content"
    >
      <section
        aria-labelledby="cast-title"
        data-testid="show-cast-section"
      >
        <ul
          class="cast-list"
        >
          <div
            data-testid="cast-list-item"
          >
            <p>
              Bryan Cranston
            </p>
            <p>
              Walter White
            </p>
          </div>
          <div
            data-testid="cast-list-item"
          >
            <p>
              Aaron Paul
            </p>
            <p>
              Jesse Pinkman
            </p>
          </div>
        </ul>
        <div
          data-testid="pagination"
        >
          <button
            disabled=""
          >
            Prev
          </button>
          <button>
            Next
          </button>
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
